client/channel/channel.go

```diff
@@ -289,23 +309,45 @@ func (c *Channel) subscribe(id uint32) chan agentpb.ServerResponsePayload {
 	return ch
 }
 
-func (c *Channel) publish(id uint32, resp agentpb.ServerResponsePayload) {
+func (c *Channel) removeResponseChannel(id uint32) (chan Response, error) {
 	c.m.Lock()
-	if c.responses == nil { // Channel is closed, no more publishing
+	if c.responses == nil { // Channel is closed
 		c.m.Unlock()
-		return
+		return nil, nil
 	}
 
 	ch := c.responses[id]
+	c.m.Unlock()
 	if ch == nil {
-		c.m.Unlock()
-		c.close(errors.WithStack(fmt.Errorf("no subscriber for ID %d", id)))
-		return
+		return nil, errors.WithStack(fmt.Errorf("no subscriber for ID %d", id))
 	}
-
+	c.m.Lock()
 	delete(c.responses, id)
 	c.m.Unlock()
-	ch <- resp
+	return ch, nil
+}
+
+// cancel sends an error to the subscriber.
+func (c *Channel) cancel(id uint32, err error) {
+	if ch, _ := c.removeResponseChannel(id); ch != nil {
+		ch <- Response{Error: err}
+	}
+}
+
+func (c *Channel) publish(id uint32, status *protostatus.Status, resp agentpb.ServerResponsePayload) {
```

Comment:
nit: lets put publish and subscribe methods next to each other so that the code is easier to understand

Reply:
Done